linkcheck/htmlutil/formsearch.py

```diff
@@ -37,60 +37,24 @@ def __repr__(self):
         return "<url=%s data=%s>" % (self.url, self.data)
 
 
-class FormFinder(object):
-    """Base class handling HTML start elements.
-    TagFinder instances are used as HtmlParser handlers."""
-
-    def __init__(self):
-        """Initialize local variables."""
-        super(FormFinder, self).__init__()
-        self.forms = []
-        self.form = None
-
-    def start_element(self, tag, attrs, element_text, lineno, column):
-        """Does nothing, override in a subclass."""
-        if tag == u'form':
-            if u'action' in attrs:
-                url = attrs['action']
-                self.form = Form(url)
-        elif tag == u'input':
-            if self.form:
-                if 'name' in attrs:
-                    key = attrs['name']
-                    value = attrs.get('value')
-                    self.form.add_value(key, value)
-                else:
-                    log.warn(LOG_CHECK, "nameless form input %s" % attrs)
-                    pass
-            else:
-                log.warn(LOG_CHECK, "formless input %s" % attrs)
-                pass
-
-    def start_end_element(self, tag, attrs, element_text, lineno, column):
-        """Delegate a combined start/end element (eg. <input .../>) to
-        the start_element method. Ignore the end element part."""
-        self.start_element(tag, attrs, element_text, lineno, column)
-
-    def end_element(self, tag):
-        """search for ending form values."""
-        if tag == u'form':
-            self.forms.append(self.form)
-            self.form = None
-
-
 def search_form(content, cgiuser, cgipassword):
     """Search for a HTML form in the given HTML content that has the given
     CGI fields. If no form is found return None.
     """
-    handler = FormFinder()
-    parser = htmlsax.parser(handler)
-    # parse
-    parser.feed_soup(htmlsax.make_soup(content))
-    log.debug(LOG_CHECK, "Found forms %s", handler.forms)
-    cginames = (cgiuser.lower(), cgipassword.lower())
-    for form in handler.forms:
-        for key, value in form.data.items():
-            if key.lower() in cginames:
-                return form
+    soup = htmlsax.make_soup(content)
+    # The value of the name attribute is case-insensitive
+    # https://www.w3.org/TR/html401/interact/forms.html#adef-name-INPUT
+    cginames = {cgiuser.lower(), cgipassword.lower()}
+    for form_element in soup.find_all("form", action=True):
+        form = Form(form_element["action"])
+        for input_element in form_element.find_all("input",
+                                                   attrs={"name": True}):
+            form.add_value(
+                input_element["name"], input_element.attrs.get("value"))
+        if cginames <= {x.lower() for x in form.data}:
+            log.debug(LOG_CHECK, "Found form %s", form)
+            return form
```

Comment:
#343 reminded me about one subtlety in this change: the old code would find and return a form that had just one of the two fields (username + password) that we're looking for.  The new code always requires both.

There are some pages out there that are password-protected without a username field (Mailman admin/moderation web UI comes to mind), so maybe we should think about supporting that.

(I gather that while the old linkchecker code might have supported that here, in search_form, it didn't actually support it fully elsewhere, in aggregator.py, where it wanted to submit the form with both fields.  Sort of like the case insensitivity that was half-supported in one place, but not supported in the other place.)

Reply:
> I gather that while the old linkchecker code might have supported that here, in search_form, it didn't actually support it fully elsewhere

I think that is the key point for now - to make what is there more robust. I've posted #364 with some more updates to the login form handling on the end, including the case sensitivity  and an additional check in aggregator.py.
To support user name and/or password (which would be a new feature) seems to need some careful thought, how do you know what the user intended so that you can raise the appropriate errors? My first guess would be to use whether user name and password have been configured to choose which fields to pass to search_form, but that needs some investigation.

Reply:
I've added #366 for how this might work (at least in Aggregate and search_form).